unitTests/test_slice.cpp

```diff
@@ -117,16 +120,21 @@ TYPED_TEST_P(slice, atAccess)
     }
 }
 
-// TODO C++11: test range based for loop
 TYPED_TEST_P(slice, iteratorAccess)
 {
     Slice<TypeParam> sl = this->getTestSlice();
 
-    std::vector<int>::const_iterator vec_it = this->vec_.begin() + 1;
+    auto vec_it = this->vec_.begin() + 1;
     for (typename Slice<TypeParam>::const_iterator it = sl.cbegin(); it < sl.cend(); ++it, ++vec_it) {
         ASSERT_EQ(*it, *vec_it);
     }
 
+    auto vec_it_second = this->vec_.begin() + 1;
+    for (const auto& val : sl) {
+        ASSERT_EQ(val, *vec_it_second);
+        ++vec_it_second;
+    }
+
     ASSERT_THROW(sl.at(sl.size()), std::out_of_range);
```

Comment:
[suggestion] When unit-testing a desirable goal is to have the minimum number of assertions possible in each test, and to have names of the tests that clearly indicate what could go wrong if the assertion fails. Line 138 is testing a corning case, and therefore I would suggest you to have it in a different test. Something like this:

```
TYPED_TEST_P(slice, throwsTryingToAccessItemOutOfBounds)
{
    Slice<TypeParam> slice = this->getTestSlice();
    ASSERT_THROW(slice.at(slice.size()), std::out_of_range);
}
```

Reply:
To be honest, I don't really like these very short unit test functions coupled with overlong `camelCaseFunctionNamesThatExplainWhatCornerCaseThisOneIsTestingBecauseWeHaveThirtyOfThese`

(Sorry for that one, I just had to do that ;-) )

Jokes aside: do you know if gtest can provide some additional form of test grouping like Python's unit test, where you can just put tests into their own modules and thereby get a nice hierarchy?
And a since we already had different opinions about this in the past: what's wrong with creating longer tests and adding some documentation instead of making the tests self-documenting? (We all know self-documenting code doesn't work, so why should it work for tests?)

Back to topic: I actually can't remember why line 138 ended up in this test, it surely doesn't belong there. I'll move it somewhere else.

Reply:
I am not aware of such test structuring in GTest, that's why some people took the direction of trying to name tests with some conventions. In my new job we are using [Catch2](https://github.com/catchorg/Catch2/blob/master/docs/test-cases-and-sections.md#top)  and I really love the fact that you can use **strings** for the test names and have SECTIONs inside the TEST_CASES.

> what's wrong with creating longer tests and adding some documentation instead of making the tests self-documenting?

If tests pass then everybody is happy 🎊 . The main argument is that if the test fail, then it takes you more mental effort to find the origin of the failure, because:

1. The test is long and you have to analyse more line about the setup part, the functions called and the assertions. 
2. In this case, the test name is `iteratorAccess` but actually line 138 is testing the `at()` method and not the iterators. That's another risk of doing more than 1 action by test ... you can end up lying about what the test is actually doing. 
3. By just reading the test name, you cannot find what could be happening. I'll try to illustrate this with an example:

```
// Current case, tests passing
[----------] 9 tests from slice/3, where TypeParam = int const * __ptr64
[ RUN      ] slice/3.atAccess
[       OK ] slice/3.atAccess (0 ms)
[ RUN      ] slice/3.iteratorAccess
[       OK ] slice/3.iteratorAccess (0 ms)

// Current case, test fail (I changed the exception type in the assertion)
[ RUN      ] slice/3.atAccess
[       OK ] slice/3.atAccess (0 ms)
[ RUN      ] slice/3.iteratorAccess  // I would expect that something about iterators is failing ... but it is actually something about the at() method!
..\unitTests\test_slice.cpp(138): error: Expected: sl.at(sl.size()) throws an exception of type std::runtime_error.
  Actual: it throws a different type.
[  FAILED  ] slice/3.iteratorAccess, where TypeParam = int const * __ptr64 (2 ms)

// Proposed test name
[ RUN      ] slice/3.atThrowsWhenAccessingIndexOutOfBounds  // I know exactly where to look
..\unitTests\test_slice.cpp(126): error: Expected: sl.at(sl.size()) throws an exception of type std::runtime_error.
  Actual: it throws a different type.
```

Please, do not take me wrong. I am not trying to impose anything here. I am just trying to share with you what I know about testing. If you think these feedback is useful, great. If not, I can live with the current version 😉